framework/decode/vulkan_address_replacer.cpp

```diff
@@ -0,0 +1,791 @@
+        {
+            auto geometry = const_cast<VkAccelerationStructureGeometryKHR*>(build_geometry_info.pGeometries != nullptr
+                                                                                ? build_geometry_info.pGeometries + j
+                                                                                : build_geometry_info.ppGeometries[j]);
+            switch (geometry->geometryType)
+            {
+                case VK_GEOMETRY_TYPE_TRIANGLES_KHR:
+                {
+                    auto& triangles = geometry->geometry.triangles;
+                    address_remap(triangles.vertexData.deviceAddress);
+                    address_remap(triangles.indexData.deviceAddress);
+                    break;
+                }
+                case VK_GEOMETRY_TYPE_AABBS_KHR:
+                {
+                    auto& aabbs = geometry->geometry.aabbs;
+                    address_remap(aabbs.data.deviceAddress);
+                    break;
+                }
+                case VK_GEOMETRY_TYPE_INSTANCES_KHR:
+                {
+                    auto& instances = geometry->geometry.instances;
+                    address_remap(instances.data.deviceAddress);
+
+                    // replace VkAccelerationStructureInstanceKHR::accelerationStructureReference inside buffer
+                    for (uint32_t k = 0; k < range_info->primitiveCount; ++k)
+                    {
+                        VkDeviceAddress accel_structure_reference =
+                            instances.data.deviceAddress + k * sizeof(VkAccelerationStructureInstanceKHR) +
+                            offsetof(VkAccelerationStructureInstanceKHR, accelerationStructureReference);
+                        addresses_to_replace.push_back(accel_structure_reference);
+                    }
+                    break;
+                }
+                default:
+                    GFXRECON_LOG_ERROR(
+                        "OverrideCmdBuildAccelerationStructuresKHR: unhandled case in switch-statement: %d",
+                        geometry->geometryType);
+                    break;
+            }
+        }
+    }
+
+    if (!addresses_to_replace.empty())
+    {
+        // prepare linear hashmap
+        hashmap_bda_.clear();
+        auto acceleration_structure_map = address_tracker.GetAccelerationStructureDeviceAddressMap();
+        for (const auto& [capture_address, replay_address] : acceleration_structure_map)
+        {
+            if (force_replace || capture_address != replay_address)
+            {
+                // store addresses we will need to replace
+                hashmap_bda_.put(capture_address, replay_address);
+            }
+        }
+
+        if (!hashmap_bda_.empty())
+        {
+            // mark injected commands
+            decode::BeginInjectedCommands();
+
+            if (pipeline_bda_ == VK_NULL_HANDLE)
+            {
+                init_pipeline();
```

Comment:
If we hit a fatal error inside this, we should abandon and jump out.

Reply:
done. we check, print an error and bail out in case this goes wrong